### PR TITLE
Improve RS_ActionDimLinear features. 

### PR DESCRIPTION
**Pull request by jacques-chen**

### CODE REVIEW

File: librecad/src/actions/rs_actiondimension.cpp [hunk 1]
```diff
@@ -91,8 +91,6 @@ bool RS_ActionDimension::isDimensionAction(RS2::ActionType type) {
 	switch(type){
 	case RS2::ActionDimAligned:
 	case RS2::ActionDimLinear:
-	case RS2::ActionDimLinearVer:
-	case RS2::ActionDimLinearHor:
```
Comment (lordofbikes) [inline librecad/src/actions/rs_actiondimension.cpp:95]: What's the reason to remove these two useful tools?

**Comment by jacques-chen** [comment]
This PR is to merge ActionDimLinear, ActionDimLinearVer, ActionDimLinearHor features, by mouse movement.



***@***.***
 
From: Armin Stebich
Date: 2023-07-28 14:40
To: LibreCAD/LibreCAD
CC: jacques-chen; Author
Subject: Re: [LibreCAD/LibreCAD] Improve RS_ActionDimLinear features. #1666 (PR #1667)
@lordofbikes commented on this pull request.


In librecad/src/actions/rs_actiondimension.cpp:
> @@ -91,8 +91,6 @@ bool RS_ActionDimension::isDimensionAction(RS2::ActionType type) {
 	switch(type){
 	case RS2::ActionDimAligned:
 	case RS2::ActionDimLinear:
-	case RS2::ActionDimLinearVer:
-	case RS2::ActionDimLinearHor:

What's the reason to remove these two useful tools?

File: librecad/src/actions/rs_actiondimension.cpp (hunk 1 above)
Comment (jacques-chen) [reply librecad/src/actions/rs_actiondimension.cpp:95]: This PR is to merge ActionDimLinear, ActionDimLinearVer, ActionDimLinearHor features, by mouse movement.

**Comment by lordofbikes** [comment]
This PR remove tools `DimLinearVer` and `DimLinearHor` completely.
Can you explain how your changes achieve to replace them?
Is their functionality still available?

**Comment by jacques-chen** [comment]
in mouseMoveEvent function,you can see

` 
   case SetDefPoint:
        if (edata->extensionPoint1.valid && edata->extensionPoint2.valid) {
            if (isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)) {
                //setAngle(0.0);
            }
            else  if (isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     !isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)){
                setAngle(0.0);
            }
            else if (!isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)) {
                setAngle(M_PI_2);
            }
            else if (!isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     !isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)){
                //setAngle(M_PI_2);
            }
`


According to the mouse coordinate point and the position of extensionPoint1 and extensionPoint2, it is currently necessary to use DimLinearVer or DimLinearHor



***@***.***
 
From: Armin Stebich
Date: 2023-07-28 15:10
To: LibreCAD/LibreCAD
CC: jacques-chen; Author
Subject: Re: [LibreCAD/LibreCAD] Improve RS_ActionDimLinear features. (PR #1667)
This PR remove tools DimLinearVer and DimLinearHor completely.
Can you explain how your changes achieve to replace them?
Is their functionality still available?
―

**Comment by jacques-chen** [comment]
> This PR remove tools `DimLinearVer` and `DimLinearHor` completely. Can you explain how your changes achieve to replace them? Is their functionality still available?

in mouseMoveEvent function,you can see

` 
   case SetDefPoint:
        if (edata->extensionPoint1.valid && edata->extensionPoint2.valid) {
            if (isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)) {
                //setAngle(0.0);
            }
            else  if (isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     !isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)){
                setAngle(0.0);
            }
            else if (!isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)) {
                setAngle(M_PI_2);
            }
            else if (!isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     !isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)){
                //setAngle(M_PI_2);
            }
`


According to the mouse coordinate point and the position of extensionPoint1 and extensionPoint2, it is currently necessary to use DimLinearVer or DimLinearHor

> in mouseMoveEvent function,you can see ` case SetDefPoint: if (edata->extensionPoint1.valid && edata->extensionPoint2.valid) { if (isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) && isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)) { //setAngle(0.0); } else if (isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) && !isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)){ setAngle(0.0); } else if (!isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) && isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)) { setAngle(M_PI_2); } else if (!isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) && !isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)){ //setAngle(M_PI_2); } ` According to the mouse coordinate point and the position of extensionPoint1 and extensionPoint2, it is currently necessary to use DimLinearVer or DimLinearHor

in mouseMoveEvent function,you can see

`
   case SetDefPoint:
        if (edata->extensionPoint1.valid && edata->extensionPoint2.valid) {
            if (isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)) {
                //setAngle(0.0);
            }
            else  if (isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     !isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)){
                setAngle(0.0);
            }
            else if (!isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)) {
                setAngle(M_PI_2);
            }
            else if (!isBetween(mouse.x, (edata->extensionPoint1).x, (edata->extensionPoint2).x) &&
                     !isBetween(mouse.y, (edata->extensionPoint1).y, (edata->extensionPoint2).y)){
                //setAngle(M_PI_2);
            }
`


According to the mouse coordinate point and the position of extensionPoint1 and extensionPoint2, it is currently necessary to use DimLinearVer or DimLinearHor

**Comment by dxli** [comment]
Conceptually, "Horizontal/Vertical" dims are also "Linear", but the difference is when you rotate them. "Linear" dimensions should be rotated along with the attached entity. "Horizontal/Vertical" will stay horizontal/vertical but the extension points be rotated along with the attached entity. Of course, since we don't have a constraint solver, features like dimension transforming are ugly.

Technical side, isBetween(x, min, max) should be robust to handle when (x, min, and max) are all equal within the tolerance.

I feel it's a good feature for "Linear", but Horizontal/Vertical dimensions should be kept, since they have special meanings for most CAD users.

**Comment by jacques-chen** [comment]
use isBetween(x, min, max) will get same effect，Horizontal/Vertical is special case of "Linear" dimensions

**Comment by dxli** [comment]
Yes, Horizontal/Vertical dimensions are also Linear, but they are still different.

For example,

A drawing has a horizontal line of a length of 10. A user may want to add both a Horizontal and a Linear dimension on it, and both dimension shows a size of 10. When the line is rotated by 45 degrees, the horizontal dim should now show a size of 7.07, but the linear dim will keep its size of 10. To the end user, those two dimensions are different types for different purposes.

If we merge the types, we can't differentiate a linear dimension happens to be horizontal and a truly horizontal.

**Comment by jacques-chen** [comment]
in this emanple we should use dim aligned

**Comment by dxli** [comment]
> in this emanple we should use dim aligned

Yes, you are right. I got confused here. I will review how other CAD apps handles dimensions.

My concern is still many CAD users relies Horizontal/Vertical dimensions. It's better to keep them, leaving the new "Linear" as an alternate.

**Comment by gerardbm** [comment]
Just to participate on this, I use a lot of CAD. A lot of times when I measure a line I need to see the length in true magnitude and in polar axis. This is how Allplan works. It shows the delta X and the delta Y:

**Comment by jacques-chen** [comment]
I had tried Horizontal/Vertical dimensions in LibreCAD.  And I give up to pull the request to improve RS_ActionDimLinear. I will fork another version.

**Comment by lordofbikes** [comment]
I recommend that you start using branches for your contributions.
See https://docs.github.com/en/get-started/quickstart/contributing-to-projects#creating-a-branch-to-work-on
By merging LibreCAD/LibreCAD `master` into your modified `master` fork at jacques-chen/LibreCAD the `master` branches diverge and possibly makes merging pull requests more complicated.

**Comment by jacques-chen** [comment]
try to open it ,you will found something wrong



***@***.***
 
From: Armin Stebich
Date: 2023-07-29 22:21
To: LibreCAD/LibreCAD
CC: jacques-chen; State change
Subject: Re: [LibreCAD/LibreCAD] Improve RS_ActionDimLinear features. (PR #1667)
I recommend that you start using branches for your contributions.
See https://docs.github.com/en/get-started/quickstart/contributing-to-projects#creating-a-branch-to-work-on
By merging LibreCAD/LibreCAD master into your modified master fork at jacques-chen/LibreCAD the master branches diverge and possibly makes merging pull requests more complicated.
―

**Pull request closed without merging**